packages/core/src/browser/shell/tab-bars.ts

```diff
@@ -74,8 +75,14 @@ export class TabBarRenderer extends TabBar.Renderer {
     // TODO refactor shell, rendered should only receive props with event handlers
     // events should be handled by clients, like ApplicationShell
     // right now it is mess: (1) client logic belong to renderer, (2) cyclic dependencies between renderes and clients
-    constructor(protected readonly contextMenuRenderer?: ContextMenuRenderer) {
+    constructor(
+        protected readonly contextMenuRenderer?: ContextMenuRenderer,
+        protected readonly decoratorService?: TabBarDecoratorService
+    ) {
         super();
+        if (this.decoratorService) {
+            this.decoratorService.onDidChangeDecorations(() => {});
```

Comment:
what is a lifecycle of `TabBarRenderer`? Is it one for application and that's why listeners should not be collected?

Generally renderers should not listen to something, but their owners should listen and call renderers when it is required. Otherwise it can lead to memory leaks.

Reply:
I understand the renderers are not supposed to include event listeners. But it seems for tab bar, it is not feasible to handle the events listeners in the owner class (`ScrollableTabBar`) and then pass the data to the renderer.

As for the lifecycle, I added the method to dispose the `decoratorService` and its event listener, which should solve the potential memory leak issue.